text/0099-otlp-http.md

```diff
@@ -0,0 +1,146 @@
+
+- gRPC is a relatively big dependency, which some clients are not willing to
+  take. Plain HTTP is a smaller dependency and is built in the standard
+  libraries of many programming languages.
+
+## OTLP/HTTP Protocol Details
+
+This proposal keeps the existing specification of OTLP over gRPC transport
+(OTLP/gRPC for short) and defines an additional way to use OTLP protocol over
+HTTP transport (OTLP/HTTP for short). OTLP/HTTP uses the same ProtoBuf payload
+that is used by OTLP/gRPC and defines how this payload is communicated over HTTP
+transport.
+
+OTLP/HTTP uses HTTP POST requests to send telemetry data from clients to
+servers. Implementations MAY use HTTP/1.1 or HTTP/2 transports. Implementations
+that use HTTP/2 transport SHOULD fallback to HTTP/1.1 transport if HTTP/2
+connection cannot be established.
+
+### Request
+
+Telemetry data is sent via HTTP POST request.
+
+The default URL path for requests that carry trace data is `/traces` (for
+example the full URL when connecting to "example.com" server will be
+`https://example.com/traces`). The request body is a ProtoBuf-encoded
+[`ExportTraceServiceRequest`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/trace/v1/trace_service.proto#L38)
+message.
+
+The default URL path for requests that carry metric data is `/metrics` and the
+request body is a ProtoBuf-encoded
+[`ExportMetricsServiceRequest`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/metrics/v1/metrics_service.proto#L35)
+message.
+
+The client MUST set "Content-Type: application/x-protobuf" request header. The
+client MAY gzip the content and in that case SHOULD include "Content-Encoding:
+gzip" request header. The client MAY include "Accept-Encoding: gzip" request
+header if it can receive gzip-encoded responses.
+
+Non-default URL paths for requests MAY be configured on the client and server
+sides.
+
+### Response
+
+#### Success
+
+On success the server MUST respond with `HTTP 200 OK`. Response body MUST be
+ProtoBuf-encoded
+[`ExportTraceServiceResponse`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/trace/v1/trace_service.proto#L47)
+message for traces and
+[`ExportMetricsServiceResponse`](https://github.com/open-telemetry/opentelemetry-proto/blob/e6c3c4a74d57f870a0d781bada02cb2b2c497d14/opentelemetry/proto/collector/metrics/v1/metrics_service.proto#L44)
+message for metrics.
+
+The server MUST set "Content-Type: application/x-protobuf" response header. If
+the request header "Accept-Encoding: gzip" is present in the request the server
+MAY gzip-encode the response and set "Content-Encoding: gzip" response header.
+
+The server SHOULD respond with success no sooner than after successfully
+decoding and validating the request.
+
+#### Failures
```

Comment:
This is specified in the catch-all "All Other Responses" section, which says - just follow HTTP spec. I do not think this protocol should prescribe any specific auth behavior. HTTP has provisions for authentication (e.g. Basic auth and 401 Unauthorized), but some backends use custom secrets in header (aka API key).

Since this is a vendor-specific behavior and there are known practices for auth in HTTP I do not want to prescribe any particular auth approach in this protocol.

In practice this means that OTLP/HTTP client implementations will have to allow users to supply any HTTP headers to be included in the request. This is what we already do for OTLP and OpenCensus exporter implementations in OpenTelemetry Collector. Perhaps the ability to include user-defined headers is worth calling out as a SHOULD requirement for clients? I am not sure.

Reply:
A general section stating "use what's normally used in HTTP" wouldn't hurt, to address such questions going forward.

Reply:
@yurishkuro I think it is already in the doc:

>All other HTTP responses that are not explicitly listed in this document should
be treated according to HTTP specification.

Does this need additional clarification?